beeline/src/java/org/apache/hive/beeline/Commands.java

```diff
@@ -1272,37 +1279,59 @@ private void debug(String message) {
     beeLine.debug(message);
   }
 
-
-
   static class LogRunnable implements Runnable {
     private final Commands commands;
     private final HiveStatement hiveStatement;
     private final long queryProgressInterval;
+    private final InPlaceUpdateStream.EventNotifier notifier;
 
     LogRunnable(Commands commands, HiveStatement hiveStatement,
-        long queryProgressInterval) {
+        long queryProgressInterval, InPlaceUpdateStream.EventNotifier eventNotifier) {
       this.hiveStatement = hiveStatement;
       this.commands = commands;
       this.queryProgressInterval = queryProgressInterval;
+      this.notifier = eventNotifier;
     }
 
-    private void updateQueryLog() throws SQLException {
-      for (String log : hiveStatement.getQueryLog()) {
-        commands.beeLine.info(log);
+    private void updateQueryLog() {
+      try {
+        List<String> queryLogs = hiveStatement.getQueryLog();
+        for (String log : queryLogs) {
+          commands.beeLine.info(log);
+        }
+        if (!queryLogs.isEmpty()) {
+          notifier.operationLogShowedToUser();
+        }
+      } catch (SQLException e) {
+        commands.error(new SQLWarning(e));
       }
     }
 
     @Override public void run() {
-      while (hiveStatement.hasMoreLogs()) {
-        try {
-          updateQueryLog();
-          Thread.sleep(queryProgressInterval);
-        } catch (SQLException e) {
-          commands.error(new SQLWarning(e));
-        } catch (InterruptedException e) {
-          commands.debug("Getting log thread is interrupted, since query is done!");
-          commands.showRemainingLogsIfAny(hiveStatement);
+      try {
+        while (!Thread.currentThread().isInterrupted()) {
+          /*
+            get the operation logs once and print it, then wait till progress bar update is complete
+            before printing the remaining logs.
+          */
+          if (notifier.canOutputOperationLogs() && hiveStatement.hasMoreLogs()) {
+            commands.debug("going to print operations logs");
+            updateQueryLog();
+            commands.debug("printed operations logs");
+          } else {
```

Comment:
the loop (!Thread.currentThread().isInterrupted()) is required as this is the code that loops for ever, the if clause conditional will be false after operations logs is printed the first time which will lead to breaking the loop and end of thread execution thus printing the results before remainder of the logs